synapse/storage/monthly_active_users.py

```diff
@@ -106,29 +107,44 @@ def _reap_users(txn):
                 # While Postgres does not require 'LIMIT', but also does not support
                 # negative LIMIT values. So there is no way to write it that both can
                 # support
-                safe_guard = self.hs.config.max_mau_value - len(self.reserved_users)
-                # Must be greater than zero for postgres
-                safe_guard = safe_guard if safe_guard > 0 else 0
-                query_args = [safe_guard]
-
-                base_sql = """
-                    DELETE FROM monthly_active_users
-                    WHERE user_id NOT IN (
-                        SELECT user_id FROM monthly_active_users
-                        ORDER BY timestamp DESC
-                        LIMIT ?
+                if len(self.reserved_users) == 0:
+                    sql = """
+                        DELETE FROM monthly_active_users
+                        WHERE user_id NOT IN (
+                            SELECT user_id FROM monthly_active_users
+                            ORDER BY timestamp DESC
+                            LIMIT ?
                         )
-                    """
+                        """
+                    txn.execute(sql, (max_mau_value,))
                 # Need if/else since 'AND user_id NOT IN ({})' fails on Postgres
                 # when len(reserved_users) == 0. Works fine on sqlite.
-                if len(self.reserved_users) > 0:
-                    query_args.extend(self.reserved_users)
-                    sql = base_sql + """ AND user_id NOT IN ({})""".format(
+                else:
+                    safe_guard = max_mau_value - len(self.reserved_users)
+                    # Must be greater than zero for postgres
+                    safe_guard = safe_guard if safe_guard > 0 else 0
```

Comment:
Is this not just `safe_guard = max(max_mau_value - len(self.reserved_users), 0)`. Also can we rename it to something like `num_users_to_remove`?

Reply:
have renamed, but the sense is the opposite, because it is number of non reserved users to save.